### ISSUE TITLE
Implement strategy for assessing the quality of the model during lifelong training

### ISSUE THREAD

**Issue by jcohenadad** [sct-pipeline/contrast-agnostic-softseg-spinalcord#87]
As we are adding more contrasts and re-training the model overtime (see eg: #83, #74, https://github.com/ivadomed/canproco/issues/46), we need to put in place a quality check assessment of model performance shift across various data domains (ie: monitor catastrophic forgetting).

**Comment by jcohenadad** [comment]
one idea of validation is to compute the CSA variation across contrasts from the test set of the spine generic data

**Comment by naga-karthik** [comment]
with each new release adding more contrasts and datasets, we observed a disturbing upward trend in absolute csa error across all the deployed models, suggesting that the drift is way too high and the model is losing in "contrast-agnostic"-ness in some sense. 

<details>
<summary> abs. csa error drift across deployed models </summary>

![abs_csa_error_perslice](https://github.com/user-attachments/assets/c071bd70-91db-4ec2-ac37-45deb94ba4c8)

</details> 

The no. 1 suspicion is the dataset imbalance created by an unusually high number of T2w and T2star images in the training set of the later models (i.e. v2.4, v2.5 etc)

Next step:

- [ ] create a balanced dataset with similar number of images for each contrast

**Comment by naga-karthik** [comment]
I created a balanced version of the aggregated dataset. Below are the details and the splits. Each contrast now has approximately 150 images (a hard-coded value after looking at how many _total_ number of images each contrast has). Except for a few contrasts (e.g. STIR) all contrasts have at least 150 images. 

<details>
<summary> balanced dataset statistics </summary>

```
SPLITS ACROSS DIFFERENT CONTRASTS (n=11):

| contrast   |   train |   validation |   test |   #images_per_contrast |
|:-----------|--------:|-------------:|-------:|-----------------------:|
| dwi        |     150 |          102 |      0 |                    252 |
| mt-off     |     146 |           97 |      0 |                    243 |
| mt-on      |     150 |           98 |      0 |                    248 |
| part-mag   |     122 |           54 |      0 |                    176 |
| psir       |     150 |           96 |      0 |                    246 |
| stir       |      57 |           32 |      0 |                     89 |
| t1map      |      61 |           27 |      0 |                     88 |
| t1w        |     150 |          127 |      0 |                    277 |
| t2star     |     150 |          150 |    116 |                    416 |
| t2w        |     150 |          150 |    150 |                    450 |
| unit1      |     132 |           59 |      0 |                    191 |
| TOTAL      |    1418 |          992 |    266 |                   2676 |


CONTRAST-WISE PATHOLOGY SPLIT (a subject can have multiple contrasts):
includes train/val/test images

|          |   MS |   HC |   RIS |   RRMS |   PPMS |   DCM |   SPMS |   SCI |   ALS |   SYR |   NMO |   #total_per_contrast |
|:---------|-----:|-----:|------:|-------:|-------:|------:|-------:|------:|------:|------:|------:|----------------------:|
| dwi      |    1 |  173 |     0 |      0 |      0 |    68 |      0 |     4 |     5 |     1 |     0 |                   252 |
| mt-off   |    0 |  184 |     0 |      0 |      0 |    59 |      0 |     0 |     0 |     0 |     0 |                   243 |
| mt-on    |    0 |  179 |     0 |      0 |      0 |    65 |      0 |     0 |     4 |     0 |     0 |                   248 |
| part-mag |    0 |    0 |     0 |    138 |     10 |     0 |     28 |     0 |     0 |     0 |     0 |                   176 |
| psir     |    0 |   29 |    40 |    146 |     31 |     0 |      0 |     0 |     0 |     0 |     0 |                   246 |
| stir     |    0 |   10 |     7 |     56 |     16 |     0 |      0 |     0 |     0 |     0 |     0 |                    89 |
| t1map    |    0 |    0 |     0 |     69 |      5 |     0 |     14 |     0 |     0 |     0 |     0 |                    88 |
| t1w      |   17 |  204 |     0 |      0 |      0 |    43 |      0 |     0 |     0 |     1 |    12 |                   277 |
| t2star   |  207 |  125 |     0 |      0 |      0 |    67 |      0 |     2 |    14 |     1 |     0 |                   416 |
| t2w      |  119 |   75 |    12 |     54 |     14 |   109 |      0 |    62 |     5 |     0 |     0 |                   450 |
| unit1    |   50 |   53 |     0 |     69 |      5 |     0 |     14 |     0 |     0 |     0 |     0 |                   191 |
| TOTAL    |  394 | 1032 |    59 |    532 |     81 |   411 |     56 |    68 |    28 |     3 |    12 |                  2676 |

```

</details> 

Next step:
- [ ] train a model on the balanced dataset, compute abs. csa error, compare with deployed models

**Comment by naga-karthik** [comment]
Something interesting happened -- it seems that an imbalance in the no. of images per contrast in the aggregated dataset is _not_ causing the CSA drift. 

<details>
<summary> abs. csa error plot deployed models vs balanced model </summary>

![abs_csa_error_perslice](https://github.com/user-attachments/assets/406f3ac2-30f3-417b-bb8f-e8e8a86e81b5)

</details>

<details>
<summary> csa error per contrast for the balanced model </summary>

![abs_error_per_contrast_balanced](https://github.com/user-attachments/assets/c554fc4c-7949-4028-b14e-d66857ed9ce1)

</details>

Note the per-contrast CSA errors are high for T2star, DWI and MToff (GRE-T1w) contrasts, but in the [previous comment](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/issues/87#issuecomment-2508729284), we see from the dataset statistics these contrasts have 150 images each in the training set.  

I am wondering if any of the new contrasts added to the training set (after the original 6 contrasts in contrast-agnostic v2.0) are negatively impacting the performance on the remaining contrasts.

**Comment by sandrinebedard** [comment]
interesting investigation! Do you have the aboslute CSA error per contrast for each model version?

**Comment by naga-karthik** [comment]
yes!

<details>
<summary> csa error per contrast for the balanced model </summary>

![abs_error_per_contrast_balanced](https://github.com/user-attachments/assets/c554fc4c-7949-4028-b14e-d66857ed9ce1)

</details>

<details>
<summary> csa error per contrast for the v20 model </summary>

![abs_error_per_contrast_v20](https://github.com/user-attachments/assets/0a1f5764-57e4-4a62-8f94-18c834844be5)

</details>

<details>
<summary> csa error per contrast for the v21 model </summary>

![abs_error_per_contrast_v21](https://github.com/user-attachments/assets/9dc6ca90-bf5a-462c-ab6f-32b2499b58d3)

</details>

<details>
<summary> csa error per contrast for the v23 model </summary>

![abs_error_per_contrast_v23](https://github.com/user-attachments/assets/b4f0381c-55ee-4fd0-b3ed-c1485ccdc2ad)

</details>

<details>
<summary> csa error per contrast for the v24 model </summary>

![abs_error_per_contrast_v24](https://github.com/user-attachments/assets/9e52acfc-a821-41ab-bd8d-8db25a205a31)

</details>

<details>
<summary> csa error per contrast for the v25 model </summary>

![abs_error_per_contrast_v25](https://github.com/user-attachments/assets/1759db82-533b-4c9c-bdbb-3dfc86561484)

</details>

**Comment by sandrinebedard** [comment]
nice! why in v2.5 DWI has a lower error than in the balanced? (3.39 mm2 vs 12.34)

**Comment by naga-karthik** [comment]
Figuring that the "dataset imbalancement" is not the only problem, I trained another model with only DCM pathology data and 1 new contrast (more like how model v2.3 was trained). 

<details>
<summary> csa error per contrast for the balanced model trained on spine-generic, UNIT1, dcm datasets </summary>

![abs_error_per_contrast_balanced](https://github.com/user-attachments/assets/14cb38a3-49b6-4cc9-9675-8f43b8c4c831)


</details>

<details>
<summary> abs csa error plot </summary>

![abs_csa_error_perslice](https://github.com/user-attachments/assets/252601de-739d-4c65-bd22-eef460f5cb92)


</details>


It turns out that the per contrast error is still high for a few contrast despite not seeing several new contrasts. This makes me wonder if adding new contrasts is _not_ the problem but maybe, in this case, too much compressed cords in the data is a problem?
